File: carkit/education/carkit_intro2av_cpp/src/perception.cpp
// Copyright 2026 CARKit maintainers
// Licensed under the Apache License, Version 2.0.

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <memory>
#include <stdexcept>
#include <string>

#include "carkit_intro2av_cpp/perception_algorithm.hpp"
#include "carkit_perception_msgs/msg/yolo_detection2_d_array.hpp"
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/camera_info.hpp"
#include "sensor_msgs/msg/compressed_image.hpp"

// CARKit learning annotation: implements the behavior described by this file's package and module.
class Intro2AvPerception : public rclcpp::Node
{
public:
  Intro2AvPerception() : Node("carkit_intro2av_cpp_perception")
  {
    image_topic_ = declare_parameter<std::string>(
      "image_topic", "/camera/camera/color/image_raw/compressed");
    camera_info_topic_ = declare_parameter<std::string>(
      "camera_info_topic", "/camera/camera/color/camera_info");
    detection_topic_ = declare_parameter<std::string>(
      "detection_2d_topic", "/yolo/detections_2d");
    preview_topic_ = declare_parameter<std::string>(
      "inference_compressed_topic", "/yolo/inference_image/compressed");
    const auto rate = std::max(0.1, declare_parameter("max_inference_rate_hz", 10.0));
    declare_parameter("minimum_confidence", 0.20);
    declare_parameter("image_size", 448);
    declare_parameter<std::string>("model_path", "");

    const auto sensor_qos = rclcpp::SensorDataQoS().keep_last(1);
    detections_pub_ = create_publisher<
      carkit_perception_msgs::msg::YoloDetection2DArray>(detection_topic_, 10);
    preview_pub_ = create_publisher<sensor_msgs::msg::CompressedImage>(
      preview_topic_, sensor_qos);
    image_sub_ = create_subscription<sensor_msgs::msg::CompressedImage>(
      image_topic_, sensor_qos,
      [this](sensor_msgs::msg::CompressedImage::ConstSharedPtr message) {
        latest_image_ = message;
        ++image_generation_;
      });
    camera_info_sub_ = create_subscription<sensor_msgs::msg::CameraInfo>(
      camera_info_topic_, sensor_qos,
      [this](sensor_msgs::msg::CameraInfo::ConstSharedPtr message) {
        camera_info_ = message;
      });
    timer_ = create_wall_timer(
      std::chrono::duration<double>(1.0 / rate), [this]() {publish_result();});
    RCLCPP_INFO(
      get_logger(),
      "Intro2AV C++ perception ready; implement perception_algorithm.cpp");
  }

private:
  carkit_intro2av_cpp::PerceptionConfig configuration() const
  {
    carkit_intro2av_cpp::PerceptionConfig config;
    config.minimum_confidence = get_parameter("minimum_confidence").as_double();
    config.image_size = get_parameter("image_size").as_int();
    config.model_path = get_parameter("model_path").as_string();
    return config;
  }

  static void validate(const carkit_intro2av_cpp::PerceptionResult & result)
  {
    for (const auto & detection : result.detections) {
      const double values[] = {
        detection.confidence, detection.bbox_x_min, detection.bbox_y_min,
        detection.bbox_x_max, detection.bbox_y_max};
      for (const double value : values) {
        if (!std::isfinite(value)) {
          throw std::runtime_error("detector returned a non-finite value");
        }
      }
      if (detection.confidence < 0.0 || detection.confidence > 1.0) {
        throw std::runtime_error("detection confidence must be in [0, 1]");
      }
    }
  }

  void process_latest_image()
  {
    try {
      auto result = carkit_intro2av_cpp::process_image(
        *latest_image_, camera_info_ ? camera_info_.get() : nullptr, configuration());
      validate(result);
      auto output = std::make_shared<carkit_perception_msgs::msg::YoloDetection2DArray>();
      output->header = latest_image_->header;
      if (camera_info_) {
        output->image_width = camera_info_->width;
        output->image_height = camera_info_->height;
      }
      output->detections = std::move(result.detections);
      output->traffic_lights = std::move(result.traffic_lights);
      latest_output_ = output;
      latest_preview_ = result.preview ? result.preview :
        std::make_shared<sensor_msgs::msg::CompressedImage>(*latest_image_);
    } catch (const std::exception & error) {
      RCLCPP_ERROR(get_logger(), "Perception algorithm failed: %s", error.what());
      latest_output_ =
        std::make_shared<carkit_perception_msgs::msg::YoloDetection2DArray>();
      latest_output_->header = latest_image_->header;
      if (camera_info_) {
        latest_output_->image_width = camera_info_->width;
        latest_output_->image_height = camera_info_->height;
      }
      latest_preview_ =
        std::make_shared<sensor_msgs::msg::CompressedImage>(*latest_image_);
    }
    processed_generation_ = image_generation_;
  }

  void publish_result()
  {
    if (!latest_image_) {
      return;
    }
    if (processed_generation_ != image_generation_) {
      process_latest_image();
    }
    if (latest_output_) {
      detections_pub_->publish(*latest_output_);
    }
    if (latest_preview_ && preview_pub_->get_subscription_count() > 0) {
      preview_pub_->publish(*latest_preview_);
    }
  }

  std::string image_topic_, camera_info_topic_, detection_topic_, preview_topic_;
  uint64_t image_generation_{0};
  uint64_t processed_generation_{static_cast<uint64_t>(-1)};
  sensor_msgs::msg::CompressedImage::ConstSharedPtr latest_image_;
  sensor_msgs::msg::CameraInfo::ConstSharedPtr camera_info_;
  sensor_msgs::msg::CompressedImage::SharedPtr latest_preview_;
  carkit_perception_msgs::msg::YoloDetection2DArray::SharedPtr latest_output_;
  rclcpp::Publisher<carkit_perception_msgs::msg::YoloDetection2DArray>::SharedPtr
    detections_pub_;
  rclcpp::Publisher<sensor_msgs::msg::CompressedImage>::SharedPtr preview_pub_;
  rclcpp::Subscription<sensor_msgs::msg::CompressedImage>::SharedPtr image_sub_;
  rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr camera_info_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Intro2AvPerception>());
  rclcpp::shutdown();
  return 0;
}

File: carkit/education/carkit_intro2av/carkit_intro2av/perception.py
#!/usr/bin/env python3

# CARKit learning annotation: implements the behavior described by this file's package and module.
"""Reference-shaped Intro2AV perception ROS node."""

import math

from carkit_perception_msgs.msg import YoloDetection2DArray
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import CameraInfo, CompressedImage

from .perception_algorithm import PerceptionConfig, process_image


class Intro2AvPerception(Node):
    """Keep ROS transport/rate handling while students own detection."""

    def __init__(self):
        super().__init__("carkit_intro2av_perception")
        self._declare_parameters()
        self.latest_image = None
        self.camera_info = None
        self.image_generation = 0
        self.processed_generation = -1
        self.latest_output = None
        self.latest_preview = None

        self.detection_publisher = self.create_publisher(
            YoloDetection2DArray,
            self._parameter("detection_2d_topic"),
            10,
        )
        self.preview_publisher = self.create_publisher(
            CompressedImage,
            self._parameter("inference_compressed_topic"),
            qos_profile_sensor_data,
        )
        self.image_subscription = self.create_subscription(
            CompressedImage,
            self._parameter("image_topic"),
            self._on_image,
            qos_profile_sensor_data,
        )
        self.camera_info_subscription = self.create_subscription(
            CameraInfo,
            self._parameter("camera_info_topic"),
            self._on_camera_info,
            qos_profile_sensor_data,
        )
        rate = max(0.1, float(self._parameter("max_inference_rate_hz")))
        self.timer = self.create_timer(1.0 / rate, self._publish_result)
        self.get_logger().info(
            "Intro2AV perception ready: compressed camera -> typed YOLO "
            "topics; implement perception_algorithm.process_image"
        )

    def _declare_parameters(self):
        """Declare the ROS parameters and their safe default values."""
        defaults = {
            "image_topic": "/camera/camera/color/image_raw/compressed",
            "camera_info_topic": "/camera/camera/color/camera_info",
            "detection_2d_topic": "/yolo/detections_2d",
            "inference_compressed_topic": "/yolo/inference_image/compressed",
            "max_inference_rate_hz": 10.0,
            "minimum_confidence": 0.20,
            "image_size": 448,
            "model_path": "",
        }
        for name, value in defaults.items():
            self.declare_parameter(name, value)

    def _parameter(self, name):
        """Return the current value of a declared ROS parameter."""
        return self.get_parameter(name).value

    def _on_image(self, message):
        self.latest_image = message
        self.image_generation += 1

    def _on_camera_info(self, message):
        self.camera_info = message

    def _configuration(self):
        """Collect current parameters into an algorithm configuration."""
        return PerceptionConfig(
            minimum_confidence=float(self._parameter("minimum_confidence")),
            image_size=int(self._parameter("image_size")),
            model_path=str(self._parameter("model_path")),
        )

    def _publish_result(self):
        if self.latest_image is None:
            return
        if self.processed_generation != self.image_generation:
            self._process_latest_image()
        if self.latest_output is not None:
            self.detection_publisher.publish(self.latest_output)
        if (
            self.latest_preview is not None
            and self.preview_publisher.get_subscription_count() > 0
        ):
            self.preview_publisher.publish(self.latest_preview)

    def _process_latest_image(self):
        image = self.latest_image
        generation = self.image_generation
        try:
            result = process_image(
                image, self.camera_info, self._configuration()
            )
            self._validate_result(result)
            output = YoloDetection2DArray()
            output.header = image.header
            if self.camera_info is not None:
                output.image_width = self.camera_info.width
                output.image_height = self.camera_info.height
            output.detections = list(result.detections)
            output.traffic_lights = list(result.traffic_lights)
            self.latest_output = output
            self.latest_preview = result.preview or image
        except Exception as error:
            self.get_logger().error(f"Perception algorithm failed: {error}")
            self.latest_output = self._empty_output(image)
            self.latest_preview = image
        self.processed_generation = generation

    def _empty_output(self, image):
        output = YoloDetection2DArray()
        output.header = image.header
        if self.camera_info is not None:
            output.image_width = self.camera_info.width
            output.image_height = self.camera_info.height
        return output

    @staticmethod
    def _validate_result(result):
        for detection in result.detections:
            values = (
                detection.confidence,
                detection.bbox_x_min,
                detection.bbox_y_min,
                detection.bbox_x_max,
                detection.bbox_y_max,
            )
            if not all(math.isfinite(value) for value in values):
                raise ValueError("detector returned a non-finite value")
            if not 0.0 <= detection.confidence <= 1.0:
                raise ValueError("detection confidence must be in [0, 1]")


def main(args=None):
    rclpy.init(args=args)
    node = Intro2AvPerception()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
